app/controllers/sessions_controller.rb

```diff
@@ -27,10 +27,21 @@ def create
   def twitter
     # raise request.env["omniauth.auth"].to_yaml
     auth = request.env["omniauth.auth"]
-    user = User.where(uid: auth["uid"]).first || User.from_twitter(auth)
-    session[:user_id] = user.id
-    flash[:notice] = "You have been logged in through Twitter."
-    redirect_back_or root_url
+    #if signed in through regular route
+    if(session[:user_id])
+      user = 
+      if User.find(session[:user_id]).update_attributes(:name => auth.info.nickname, :password => SecureRandom.hex)
```

Comment:
We also need to update the `user.uid` as well!  And we DEFINITELY don't want to update their password since they've already set one.  Cleanly done:

```
    if current_user
      current_user.update_attributes(uid: auth.uid, name: auth.info.nickname)
      flash[:notice] = "Your Twitter account has been linked."
      redirect_to settings_path
```